asserts/validation_set.go

```diff
@@ -0,0 +1,262 @@
+	}
+	return Presence(presence), nil
+}
+
+// ValidationSetSnap holds the details about a snap constrained by a validation-set assertion.
+type ValidationSetSnap struct {
+	Name   string
+	SnapID string
+
+	Presence Presence
+
+	Revision int
+}
+
+// SnapName implements naming.SnapRef.
+func (s *ValidationSetSnap) SnapName() string {
+	return s.Name
+}
+
+// ID implements naming.SnapRef.
+func (s *ValidationSetSnap) ID() string {
+	return s.SnapID
+}
+
+func checkValidationSetSnap(snap map[string]interface{}) (*ValidationSetSnap, error) {
+	name, err := checkNotEmptyStringWhat(snap, "name", "of snap")
+	if err != nil {
+		return nil, err
+	}
+	if err := naming.ValidateSnap(name); err != nil {
+		return nil, fmt.Errorf("invalid snap name %q", name)
+	}
+
+	what := fmt.Sprintf("of snap %q", name)
+
+	snapID, err := checkStringMatchesWhat(snap, "id", what, naming.ValidSnapID)
+	if err != nil {
+		return nil, err
+	}
+
+	presence, err := checkPresence(snap, what, validValidationSetSnapPresences)
+	if err != nil {
+		return nil, err
+	}
+
+	var snapRevision int
+	if _, ok := snap["revision"]; ok {
+		var err error
+		snapRevision, err = checkSnapRevisionWhat(snap, "revision", what)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if snapRevision != 0 && presence == PresenceInvalid {
+		return nil, fmt.Errorf(`cannot specify revision %s at the same time as stating its presence is invalid`, what)
+	}
+
+	return &ValidationSetSnap{
+		Name:     name,
+		SnapID:   snapID,
+		Presence: presence,
+		Revision: snapRevision,
+	}, nil
+}
+
+func checkValidationSetSnaps(snapList interface{}) ([]*ValidationSetSnap, error) {
+	const wrongHeaderType = `"snaps" header must be a list of maps`
+
+	entries, ok := snapList.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf(wrongHeaderType)
+	}
+
+	seen := make(map[string]bool, len(entries))
+	seenIDs := make(map[string]string, len(entries))
+	snaps := make([]*ValidationSetSnap, 0, len(entries))
+	for _, entry := range entries {
+		snap, ok := entry.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf(wrongHeaderType)
+		}
+		valSetSnap, err := checkValidationSetSnap(snap)
+		if err != nil {
+			return nil, err
+		}
+
+		if seen[valSetSnap.Name] {
+			return nil, fmt.Errorf("cannot list the same snap %q multiple times", valSetSnap.Name)
+		}
+		seen[valSetSnap.Name] = true
+		snapID := valSetSnap.SnapID
+		if underName := seenIDs[snapID]; underName != "" {
+			return nil, fmt.Errorf("cannot specify the same snap id %q multiple times, specified for snaps %q and %q", snapID, underName, valSetSnap.Name)
+		}
+		seenIDs[snapID] = valSetSnap.Name
+
+		if valSetSnap.Presence == "" {
+			valSetSnap.Presence = PresenceRequired
+		}
+
+		snaps = append(snaps, valSetSnap)
+	}
+
+	return snaps, nil
+}
+
+// ValidationSet holds a validation-set assertion, which is a
+// statement by an account about a set snaps and possibly revisions
+// for which an extrisinc/implied property is valid (e.g. they work
```

Comment:
s/extrisinc/extrinsic/

Reply:
I wonder why run-checks didn't catch that